File: game_rs/src/flappy_scene.rs
//! The whole game scene — mirror of `Flappy.FlappyScene` in C#.

use macroquad::prelude::*;

use mini_engine::{EngineContext, Scene};

use crate::bird::Bird;
use crate::config;
use crate::pipe::Pipe;
use crate::primitives::{PackedColor, Vec2};

// ---------------------------------------------------------------------------
// Colour palette (kept as PackedColor for consistency with the C# port)
// ---------------------------------------------------------------------------

const COLOR_SKY: PackedColor = PackedColor::new(115, 190, 240, 255);
const COLOR_SKY_DEEP: PackedColor = PackedColor::new(100, 175, 235, 255);
const COLOR_WHITE: PackedColor = PackedColor::new(255, 255, 255, 255);
const COLOR_BLACK: PackedColor = PackedColor::new(0, 0, 0, 255);
const COLOR_GOLD: PackedColor = PackedColor::new(255, 205, 40, 255);
const COLOR_RED: PackedColor = PackedColor::new(225, 60, 55, 255);
const COLOR_DARK_GREEN: PackedColor = PackedColor::new(40, 130, 40, 255);
const COLOR_GROUND: PackedColor = PackedColor::new(220, 190, 105, 255);
const COLOR_GROUND_GRASS: PackedColor = PackedColor::new(125, 200, 90, 255);

// ---------------------------------------------------------------------------
// State machine
// ---------------------------------------------------------------------------

enum State {
    Ready,
    Playing,
    GameOver,
}

// ---------------------------------------------------------------------------
// The scene
// ---------------------------------------------------------------------------

/// The whole game: a single scene that runs the Ready → Playing → GameOver
/// state machine, spawns pipes, and handles scoring and collisions.
pub struct FlappyScene {
    state: State,
    bird: Bird,
    pipes: Vec<Pipe>,
    spawn_timer: f32,
    score: u32,
    best: u32,
}

impl FlappyScene {
    pub fn new() -> Self {
        let screen_h = screen_height();
        Self {
            state: State::Ready,
            bird: Bird::new(Vec2::new(config::BIRD_X, screen_h * 0.45)),
            pipes: Vec::new(),
            spawn_timer: 0.0,
            score: 0,
            best: 0,
        }
    }

    fn floor_y(&self, screen_h: f32) -> f32 {
        screen_h - config::GROUND_HEIGHT
    }

    fn reset(&mut self, screen_h: f32) {
        self.state = State::Ready;
        self.bird = Bird::new(Vec2::new(config::BIRD_X, screen_h * 0.45));
        self.pipes.clear();
        self.spawn_timer = 0.0;
        self.score = 0;
    }

    fn spawn_pipe(&mut self, screen_w: f32, screen_h: f32) {
        let gap_y = macroquad::rand::gen_range(config::GAP_MARGIN, self.floor_y(screen_h) - config::GAP_MARGIN);
        self.pipes.push(Pipe::new(screen_w, gap_y));
    }

    fn update_playing(&mut self, dt: f32, screen_w: f32, screen_h: f32) {
        // Spawn pipes on a timer.
        self.spawn_timer += dt;
        if self.spawn_timer >= config::SPAWN_INTERVAL {
            self.spawn_timer -= config::SPAWN_INTERVAL;
            self.spawn_pipe(screen_w, screen_h);
        }

        // Scroll and reap.
        for pipe in &mut self.pipes {
            pipe.update(dt);
        }
        self.pipes.retain(|p| !p.is_off_screen());

        // Scoring + collision.
        let bird_box = self.bird.bounds();
        for pipe in &mut self.pipes {
            if !pipe.scored && pipe.x + pipe.width < self.bird.pos.x {
                pipe.scored = true;
                self.score += 1;
            }

            let (top, bottom) = pipe.solids(screen_h);
            if top.overlaps(&bird_box) || bottom.overlaps(&bird_box) {
                self.bird.dead = true;
            }
        }

        if self.bird.dead {
            self.best = self.best.max(self.score);
            self.state = State::GameOver;
        }
    }

    // --- presentation helpers ----------------------------------------------

    fn draw_background(&self, screen_w: f32, screen_h: f32) {
        clear_background(COLOR_SKY.to_mq());
        draw_rectangle(0.0, 0.0, screen_w, screen_h * 0.55, COLOR_SKY_DEEP.to_mq());

        let clouds = [(90.0, 120.0, 26.0), (340.0, 80.0, 20.0), (250.0, 200.0, 18.0)];
        let cloud_color = COLOR_WHITE.with_alpha(217).to_mq();
        for (cx, cy, r) in clouds {
            draw_circle(cx, cy, r, cloud_color);
            draw_circle(cx + r, cy + 4.0, r * 0.8, cloud_color);
            draw_circle(cx - r, cy + 6.0, r * 0.7, cloud_color);
        }
    }

    fn draw_ground(&self, screen_w: f32, screen_h: f32) {
        let top = screen_h - config::GROUND_HEIGHT;
        draw_rectangle(0.0, top, screen_w, config::GROUND_HEIGHT, COLOR_GROUND.to_mq());
        draw_rectangle(0.0, top, screen_w, 14.0, COLOR_GROUND_GRASS.to_mq());
        draw_line(0.0, top, screen_w, top, 4.0, COLOR_DARK_GREEN.to_mq());
    }

    fn draw_hud(&self, screen_w: f32, screen_h: f32) {
        // Score, centred near the top.
        let score_text = self.score.to_string();
        let sw = measure_text(&score_text, None, 64, 1.0).width;
        let sx = (screen_w - sw) / 2.0;
        draw_text(&score_text, sx + 2.0, 84.0, 64.0, COLOR_BLACK.with_alpha(100).to_mq());
        draw_text(&score_text, sx, 82.0, 64.0, COLOR_WHITE.to_mq());

        match self.state {
            State::Ready => {
                self.center_text("FLAPPY BIRD", 0.30, 56.0, COLOR_GOLD, screen_w, screen_h);
                self.center_text("Press SPACE / click to flap", 0.42, 26.0, COLOR_WHITE, screen_w, screen_h);
                self.center_text("ESC to quit", 0.47, 22.0, COLOR_WHITE, screen_w, screen_h);
            }
            State::GameOver => {
                draw_rectangle(0.0, 0.0, screen_w, screen_h, COLOR_BLACK.with_alpha(90).to_mq());
                self.center_text("GAME OVER", 0.32, 54.0, COLOR_RED, screen_w, screen_h);
                self.center_text(&format!("Score: {}", self.score), 0.44, 32.0, COLOR_WHITE, screen_w, screen_h);
                self.center_text(&format!("Best: {}", self.best), 0.50, 28.0, COLOR_GOLD, screen_w, screen_h);
                self.center_text("Press SPACE / click to retry", 0.60, 24.0, COLOR_WHITE, screen_w, screen_h);
            }
            State::Playing => { /* no overlay */ }
        }
    }

    fn center_text(&self, text: &str, y_frac: f32, size: f32, color: PackedColor, screen_w: f32, screen_h: f32) {
        let x = (screen_w - measure_text(text, None, size as u16, 1.0).width) / 2.0;
        let y = screen_h * y_frac;
        draw_text(text, x + 2.0, y + 2.0, size, COLOR_BLACK.with_alpha(115).to_mq());
        draw_text(text, x, y, size, color.to_mq());
    }
}

impl Scene for FlappyScene {
    fn update(&mut self, ctx: &mut EngineContext) {
        let screen_w = ctx.screen_width();
        let screen_h = ctx.screen_height();

        // Global quit key.
        if ctx.key_pressed(KeyCode::Escape) {
            ctx.quit();
            return;
        }

        let flap_input = ctx.key_pressed(KeyCode::Space)
            || ctx.key_pressed(KeyCode::Up)
            || ctx.mouse_pressed(MouseButton::Left);

        match self.state {
            State::Ready => {
                // Gentle hover until the first flap.
                let base_y = screen_h * 0.45;
                self.bird.pos = Vec2::new(self.bird.pos.x, base_y + (ctx.time as f32 * 3.0).sin() * 10.0);
                if flap_input {
                    self.state = State::Playing;
                    self.bird.flap();
                }
            }
            State::Playing => {
                self.bird.update(ctx.dt, screen_h, flap_input);
                self.update_playing(ctx.dt, screen_w, screen_h);
            }
            State::GameOver => {
                // Let the bird drop onto the ground.
                if self.bird.pos.y + self.bird.radius < self.floor_y(screen_h) {
                    self.bird.vel += config::GRAVITY * ctx.dt;
                    self.bird.pos = Vec2::new(
                        self.bird.pos.x,
                        self.bird.pos.y + self.bird.vel * ctx.dt,
                    );
                }
                if flap_input {
                    self.reset(screen_h);
                }
            }
        }
    }

    fn draw(&self, _ctx: &EngineContext) {
        let screen_w = screen_width();
        let screen_h = screen_height();

        self.draw_background(screen_w, screen_h);

        for pipe in &self.pipes {
            pipe.draw(screen_h);
        }

        self.draw_ground(screen_w, screen_h);
        self.bird.draw();
        self.draw_hud(screen_w, screen_h);
    }
}

File: dll_stresstesting/game_cs/src/FlappyScene.cs
namespace Flappy;

/// <summary>
/// The whole game: a single scene that runs the Ready -> Playing -> GameOver
/// state machine, spawns pipes, and handles scoring and collisions.
/// </summary>
public sealed class FlappyScene
{
    private enum State { Ready, Playing, GameOver }

    private State _state = State.Ready;
    private Bird _bird;
    private readonly List<Pipe> _pipes = new();
    private float _spawnTimer;
    private int _score;
    private int _best;

    public FlappyScene()
    {
        _bird = new Bird(new Vec2(Config.BirdX, 720f * 0.45f));
    }

    private float FloorY => Engine.Height - Config.GroundHeight;

    private void Reset()
    {
        _state = State.Ready;
        _bird = new Bird(new Vec2(Config.BirdX, Engine.Height * 0.45f));
        _pipes.Clear();
        _spawnTimer = 0f;
        _score = 0;
    }

    private void SpawnPipe()
    {
        float gapY = Engine.RandRange(Config.GapMargin, FloorY - Config.GapMargin);
        _pipes.Add(new Pipe(Engine.Width, gapY));
    }

    public void Update(float dt)
    {
        if (Engine.KeyPressed(Key.Escape))
        {
            Engine.Quit();
            return;
        }

        bool flapped = Engine.KeyPressed(Key.Space) || Engine.KeyPressed(Key.Up) ||
                       Engine.MouseLeftPressed();

        switch (_state)
        {
            case State.Ready:
                // Gentle hover until the first flap.
                float baseY = Engine.Height * 0.45f;
                _bird.Pos = new Vec2(_bird.Pos.X, baseY + MathF.Sin((float)Engine.Time * 3f) * 10f);
                if (flapped)
                {
                    _state = State.Playing;
                    _bird.Flap();
                }
                break;

            case State.Playing:
                UpdatePlaying(dt);
                break;

            case State.GameOver:
                // Let the bird drop onto the ground.
                if (_bird.Pos.Y + _bird.Radius < FloorY)
                {
                    _bird.Vel += Config.Gravity * dt;
                    _bird.Pos = new Vec2(_bird.Pos.X, _bird.Pos.Y + _bird.Vel * dt);
                }
                if (flapped)
                    Reset();
                break;
        }
    }

    private void UpdatePlaying(float dt)
    {
        _bird.Update(dt);

        // Spawn pipes on a timer.
        _spawnTimer += dt;
        if (_spawnTimer >= Config.SpawnInterval)
        {
            _spawnTimer -= Config.SpawnInterval;
            SpawnPipe();
        }

        // Scroll and reap.
        foreach (var pipe in _pipes)
            pipe.Update(dt);
        _pipes.RemoveAll(p => !p.IsAlive);

        // Scoring + collision.
        float screenH = Engine.Height;
        Rect birdBox = _bird.Bounds!.Value;
        foreach (var pipe in _pipes)
        {
            if (!pipe.Scored && pipe.X + pipe.Width < _bird.Pos.X)
            {
                pipe.Scored = true;
                _score++;
            }

            var (top, bottom) = pipe.Solids(screenH);
            if (top.Overlaps(birdBox) || bottom.Overlaps(birdBox))
                _bird.Dead = true;
        }

        if (_bird.Dead)
        {
            _best = Math.Max(_best, _score);
            _state = State.GameOver;
        }
    }

    public void Draw()
    {
        DrawBackground();

        foreach (var pipe in _pipes)
            pipe.Draw();

        DrawGround();
        _bird.Draw();

        DrawHud();
    }

    // --- presentation ------------------------------------------------------

    private static void DrawBackground()
    {
        float w = Engine.Width;
        float h = Engine.Height;
        Engine.ClearBackground(Color.Sky);
        Engine.DrawRectangle(0, 0, w, h * 0.55f, Color.SkyDeep);

        (float cx, float cy, float r)[] clouds =
        {
            (90, 120, 26), (340, 80, 20), (250, 200, 18),
        };
        var cloud = Color.White.WithAlpha(217);
        foreach (var (cx, cy, r) in clouds)
        {
            Engine.DrawCircle(cx, cy, r, cloud);
            Engine.DrawCircle(cx + r, cy + 4, r * 0.8f, cloud);
            Engine.DrawCircle(cx - r, cy + 6, r * 0.7f, cloud);
        }
    }

    private static void DrawGround()
    {
        float w = Engine.Width;
        float top = Engine.Height - Config.GroundHeight;
        Engine.DrawRectangle(0, top, w, Config.GroundHeight, Color.Ground);
        Engine.DrawRectangle(0, top, w, 14f, Color.GroundGrass);
        Engine.DrawLine(0, top, w, top, 4f, Color.DarkGreen);
    }

    private void DrawHud()
    {
        float w = Engine.Width;

        // Score, centred near the top.
        string score = _score.ToString();
        float sw = Engine.MeasureTextWidth(score, 64f);
        float sx = (w - sw) / 2f;
        Engine.DrawText(score, sx + 2f, 84f, 64f, Color.Black.WithAlpha(100));
        Engine.DrawText(score, sx, 82f, 64f, Color.White);

        switch (_state)
        {
            case State.Ready:
                CenterText("FLAPPY BIRD", 0.30f, 56f, Color.Gold);
                CenterText("Press SPACE / click to flap", 0.42f, 26f, Color.White);
                CenterText("ESC to quit", 0.47f, 22f, Color.White);
                break;

            case State.GameOver:
                Engine.DrawRectangle(0, 0, w, Engine.Height, Color.Black.WithAlpha(90));
                CenterText("GAME OVER", 0.32f, 54f, Color.Red);
                CenterText($"Score: {_score}", 0.44f, 32f, Color.White);
                CenterText($"Best: {_best}", 0.50f, 28f, Color.Gold);
                CenterText("Press SPACE / click to retry", 0.60f, 24f, Color.White);
                break;
        }
    }

    private static void CenterText(string text, float yFrac, float size, Color color)
    {
        float x = (Engine.Width - Engine.MeasureTextWidth(text, size)) / 2f;
        float y = Engine.Height * yFrac;
        Engine.DrawText(text, x + 2f, y + 2f, size, Color.Black.WithAlpha(115));
        Engine.DrawText(text, x, y, size, color);
    }
}
